core/plugins/org.polarsys.capella.core.sirius.analysis/src/org/polarsys/capella/core/sirius/analysis/FaServices.java

```diff
@@ -3046,13 +3061,69 @@ EObject removeUselessPortRealizations(Port port, boolean includeFunctionalRealiz
         }
 
         if (topDelegation && CsServices.getService().getContainersOfParts(containingPort).contains(delegatedPort)) {
-          elements.add(exchange);
-        } else if (bottomDelegation
-            && CsServices.getService().getContainersOfParts(delegatedPort).contains(containingPort)) {
-          elements.add(exchange);
+          delegations.add(exchange);
+        } else if (bottomDelegation && CsServices.getService().getContainersOfParts(delegatedPort).contains(containingPort)) {
+          delegations.add(exchange);
         }
       }
     }
+    return delegations;
+  }
+
+  /**
+   * This method retrieve all the port's allocations and return those that should be deleted (as they do not make sense
+   * for the new element)
+   * 
+   * @param movingPort
+   *          the port that is being D&Ded
+   * @param newContainer
+   *          the container the port was just Dropped on.
+   * @param includeFunctionalRealization
+   * @param includeComponentRealization
+   * @param topDelegation
+   * @param bottomDelegation
+   * @return
+   */
+  private Collection<Allocation> retrieveObsoletePortAllocations(Port movingPort, NamedElement newContainer, boolean includeFunctionalRealization,
+      boolean includeComponentRealization, boolean topDelegation, boolean bottomDelegation) {
+    Collection<Allocation> elements = new HashSet<Allocation>();
+    Collection<Allocation> allocations = retrievePortAllocations(movingPort, includeFunctionalRealization, includeComponentRealization);
+    Component instanciatedComponent = null;
+    if (newContainer instanceof Part) {
+      instanciatedComponent = PartExt.getComponentOfPart((Part) newContainer);
+    }
+    for (Allocation allocation : allocations) {
+      Port delegatedPort = (allocation.getSourceElement() == movingPort) ? (Port) allocation.getTargetElement() : (Port) allocation.getSourceElement();
+      if (instanciatedComponent == null || !isEObjectInHierarchyOfContainer(delegatedPort.eContainer(), instanciatedComponent)) {
+          elements.add(allocation);
+        }
+    }
+    return elements;
+  }
+
+  /**
+   * Test two objects to determine if the first one is a subElement of the other. Should eObject be an AbstractFunction,
+   * the test will be done on each of its FunctionalAllocation instead.
+   * 
+   * @param eObject
+   * @param container
+   * @return
+   */
+  private boolean isEObjectInHierarchyOfContainer(EObject eObject, NamedElement container) {
+    if (eObject == container) {
+      return true;
+    } else if (eObject instanceof AbstractFunction) {
+      if (((AbstractFunction) eObject).getComponentFunctionalAllocations().stream().anyMatch(cfa -> isEObjectInHierarchyOfContainer(cfa, container))) {
+        return true;
+      }
+    } else if (eObject instanceof ModelElement && eObject.eContainer() != null) {
+      return isEObjectInHierarchyOfContainer(eObject.eContainer(), container);

```

Comment:
Components are not always contained by eContainer feature.
For intance, you can have a Component "Wheel" that is used in several components, or several times.
![image](https://github.com/eclipse/capella/assets/34767327/c920f151-3a30-4842-b4e8-84fe0bdb93e5)

====

![image](https://github.com/eclipse/capella/assets/34767327/67779dbf-2aa2-4460-b24b-033bdd448404)

![image](https://github.com/eclipse/capella/assets/34767327/73c173c1-ac5f-493f-9b4e-76657c503dbc)

You can then create components that are located outside the parent component;

![image](https://github.com/eclipse/capella/assets/34767327/c3dcf69f-4d7c-4808-bbba-11fa784613a7)

Instead of looking for Econtainer, we need to look for owning components that are owning the parts of the component Wheel (here, bike or tribike)

(eObject + (eObject is function ? fct.getAllocationBlocks())).anyMatch( e => ComponentExt.getAllSubUsedComponents(container).contains(e) ) or 

or container == e || ComponentExt.isAncestor(container, e)